{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bioluigi.tasks as tasks\n",
    "from bioluigi.slurm import SlurmExecutableTask\n",
    "from bioluigi.decorators import requires, inherits\n",
    "from bioluigi.utils import CheckTargetNonEmpty\n",
    "import luigi\n",
    "import os,re\n",
    "from luigi import LocalTarget, Parameter\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import luigi.tools.deps_tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "maker_opts_1={\n",
    "'organism_type':'eukaryotic', \n",
    "'model_org':'fungi',\n",
    "'repeat_protein':'/tgac/software/testing/maker/2.31.8/x86_64/data/te_proteins.fasta', \n",
    "'est2genome':1, \n",
    "'protein2genome':1, \n",
    "'TMP':'/tmp'}\n",
    "\n",
    "maker_opts_2={\n",
    "'organism_type':'eukaryotic', \n",
    "'model_org':'fungi',\n",
    "'repeat_protein':'/tgac/software/testing/maker/2.31.8/x86_64/data/te_proteins.fasta', \n",
    "'est2genome':0, \n",
    "'protein2genome':0, \n",
    "'TMP':'/tmp',\n",
    "'est_pass':1,\n",
    "'protein_pass':1,\n",
    "'rm_pass':1}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference='/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/reference/Pst_104E_v13_p_ctg.fa'\n",
    "base_dir=\"/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data\"\n",
    "\n",
    "class ReferenceFasta(tasks.ReferenceFasta):\n",
    "    pass\n",
    "\n",
    "@requires(ReferenceFasta)\n",
    "class AddGenome(tasks.jbrowse.AddGenome):\n",
    "    pass\n",
    "\n",
    "@requires(ReferenceFasta)\n",
    "class RepeatMasker(tasks.repeatmodeler.RepeatMasker):\n",
    "    pass\n",
    "\n",
    "class StringTieMerge(luigi.ExternalTask):\n",
    "    def output(self):\n",
    "        return LocalTarget(\"/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data/stringtie.gff3\")\n",
    "    \n",
    "class UniProt(luigi.ExternalTask):\n",
    "    def output(self):\n",
    "        return LocalTarget(\"/usr/users/JIC_a1/buntingd/GenomeAnnotation/uniprot_PST78.fasta\")\n",
    "\n",
    "    \n",
    "## ----------- Inital MAKER run ------------- ##\n",
    "\n",
    "@inherits(ReferenceFasta,StringTieMerge,UniProt,tasks.maker.MAKER)\n",
    "class MAKER1(luigi.WrapperTask):\n",
    "    maker_opts, maker_prefix=None, None\n",
    "    def requires(self):\n",
    "        return self.clone(requires(genome=ReferenceFasta, \n",
    "                                   est_gff=StringTieMerge, \n",
    "                                   protein=UniProt)(tasks.maker.MAKER),\n",
    "                          maker_opts=maker_opts_1,\n",
    "                          maker_prefix=\"Pst_104E_v13_p_ctg__round1\")\n",
    "    def output(self):\n",
    "        return self.requires().output()\n",
    "\n",
    "@requires(MAKER1)\n",
    "class GFFMerge1(tasks.maker.GFFMerge):\n",
    "    pass\n",
    "\n",
    "    \n",
    "@requires(gff=GFFMerge1, genome=AddGenome)\n",
    "class AddGFF1(tasks.jbrowse.AddGFF):\n",
    "    pass\n",
    "\n",
    "@requires(MAKER1)\n",
    "class Maker2Jbrowse1(tasks.maker.Maker2Jbrowse):\n",
    "    pass   \n",
    "\n",
    "##----------- SNAP Training -------------##\n",
    "\n",
    "@requires(MAKER1)\n",
    "class Maker2ZFF1(tasks.maker.Maker2ZFF):\n",
    "    pass\n",
    "\n",
    "@requires(Maker2ZFF1)\n",
    "class TrainSNAP1(tasks.maker.TrainSNAP):\n",
    "    pass\n",
    "\n",
    "@requires(Maker2ZFF1)\n",
    "class FathomStats1(tasks.maker.FathomStats):\n",
    "    pass\n",
    "\n",
    "##----------- Genemark Training -------------##\n",
    "\n",
    "@requires(RepeatMasker)\n",
    "class GenemarkESTrain(tasks.genemark.GenemarkESTrain):\n",
    "    pass\n",
    "\n",
    "##----------- Second MAKER -------------##\n",
    "\n",
    "@inherits(TrainSNAP1, GenemarkESTrain)\n",
    "class MAKER2(luigi.WrapperTask):\n",
    "    maker_opts, maker_prefix=None, None\n",
    "    def requires(self):\n",
    "        return self.clone(requires(genome=ReferenceFasta, \n",
    "                                   maker_gff=GFFMerge1, \n",
    "                                   snaphmm=TrainSNAP1,\n",
    "                                   gmhmm=GenemarkESTrain)(tasks.maker.MAKER),\n",
    "                          maker_opts=maker_opts_2,\n",
    "                          maker_prefix=\"Pst_104E_v13_p_ctg_round2\")    \n",
    "    def output(self):\n",
    "        return self.requires().output()   \n",
    "\n",
    "@requires(MAKER2)\n",
    "class GFFMerge2(tasks.maker.GFFMerge):\n",
    "    pass\n",
    "\n",
    "@requires(gff=GFFMerge2, genome=AddGenome)\n",
    "class AddGFF2(tasks.jbrowse.AddGFF):\n",
    "    pass\n",
    "\n",
    "@requires(MAKER2)\n",
    "class Maker2Jbrowse2(tasks.maker.Maker2Jbrowse):\n",
    "    pass   \n",
    "\n",
    "##----------- Second SNAP Training -------------##\n",
    "\n",
    "@requires(MAKER2)\n",
    "class Maker2ZFF2(tasks.maker.Maker2ZFF):\n",
    "    pass\n",
    "\n",
    "@requires(Maker2ZFF2)\n",
    "class TrainSNAP2(tasks.maker.TrainSNAP):\n",
    "    pass\n",
    "\n",
    "@requires(Maker2ZFF2)\n",
    "class FathomStats2(tasks.maker.FathomStats):\n",
    "    pass\n",
    "\n",
    "##----------- Third MAKER -------------##\n",
    "\n",
    "@inherits(MAKER2, TrainSNAP2)\n",
    "class MAKER3(luigi.WrapperTask):\n",
    "    maker_opts, maker_prefix=None, None\n",
    "    def requires(self):\n",
    "        return self.clone(requires(genome=ReferenceFasta, \n",
    "                                   maker_gff=GFFMerge2, \n",
    "                                   snaphmm=TrainSNAP2,\n",
    "                                   gmhmm=GenemarkESTrain)(tasks.maker.MAKER),\n",
    "                          maker_opts=maker_opts_2,\n",
    "                          maker_prefix=\"Pst_104E_v13_p_ctg_round3\")    \n",
    "    def output(self):\n",
    "        return self.requires().output()   \n",
    "    \n",
    "@requires(MAKER3)\n",
    "class GFFMerge3(tasks.maker.GFFMerge):\n",
    "    pass\n",
    "\n",
    "@requires(gff=GFFMerge3, genome=AddGenome)\n",
    "class AddGFF3(tasks.jbrowse.AddGFF):\n",
    "    pass\n",
    "\n",
    "@requires(MAKER3)\n",
    "class Maker2Jbrowse3(tasks.maker.Maker2Jbrowse):\n",
    "    pass   \n",
    "\n",
    "\n",
    "##----------- Pipeline Wrapper -------------##\n",
    "@requires(Maker2Jbrowse1,Maker2Jbrowse2, Maker2Jbrowse3, \n",
    "          AddGFF1, AddGFF2, AddGFF3, \n",
    "          FathomStats1, FathomStats2)\n",
    "class MAKERPipelineWrapper(luigi.WrapperTask):\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO: Informed scheduler that task   MAKERPipelineWrapper__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d   has status   PENDING\n",
      "INFO: Informed scheduler that task   FathomStats2__usr_users_JIC_a__usr_users_JIC_a_2b875e6c10   has status   DONE\n",
      "INFO: Informed scheduler that task   FathomStats1__usr_users_JIC_a__usr_users_JIC_a_2b875e6c10   has status   DONE\n",
      "INFO: Informed scheduler that task   AddGFF3__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d   has status   PENDING\n",
      "INFO: Informed scheduler that task   GFFMerge3__usr_users_JIC_a__usr_users_JIC_a_2b875e6c10   has status   DONE\n",
      "INFO: Informed scheduler that task   AddGenome__usr_users_JIC_a__usr_users_JIC_a_5f48dcba3f   has status   PENDING\n",
      "INFO: Informed scheduler that task   ReferenceFasta__usr_users_JIC_a_2c99a42331   has status   DONE\n",
      "INFO: Informed scheduler that task   AddGFF2__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d   has status   PENDING\n",
      "INFO: Informed scheduler that task   GFFMerge2__usr_users_JIC_a__usr_users_JIC_a_2b875e6c10   has status   DONE\n",
      "INFO: Informed scheduler that task   AddGFF1__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d   has status   PENDING\n",
      "INFO: Informed scheduler that task   GFFMerge1__usr_users_JIC_a__usr_users_JIC_a_2b875e6c10   has status   DONE\n",
      "INFO: Informed scheduler that task   Maker2Jbrowse3__usr_users_JIC_a__usr_users_JIC_a_2b875e6c10   has status   DONE\n",
      "INFO: Informed scheduler that task   Maker2Jbrowse2__usr_users_JIC_a__usr_users_JIC_a_2b875e6c10   has status   DONE\n",
      "INFO: Informed scheduler that task   Maker2Jbrowse1__usr_users_JIC_a__usr_users_JIC_a_2b875e6c10   has status   DONE\n",
      "INFO: Done scheduling tasks\n",
      "INFO: Running Worker with 25 processes\n",
      "INFO: [pid 15253] Worker Worker(salt=476501024, workers=25, host=n128n20.nbicluster, username=buntingd, pid=14220) running   AddGenome(reference=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/reference/Pst_104E_v13_p_ctg.fa, jbrowse_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0)\n",
      "INFO: Tmp dir: /nbi/scratch/buntingd/AddGenome__usr_users_JIC_a__usr_users_JIC_a_5f48dcba3f\n",
      "INFO: SLURM: jobid=15034871\n",
      "INFO: \n",
      "SLURM err AddGenome__usr_users_JIC_a__usr_users_JIC_a_5f48dcba3f: \n",
      "SLURM out AddGenome__usr_users_JIC_a__usr_users_JIC_a_5f48dcba3f: \n",
      "INFO: [pid 15253] Worker Worker(salt=476501024, workers=25, host=n128n20.nbicluster, username=buntingd, pid=14220) done      AddGenome(reference=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/reference/Pst_104E_v13_p_ctg.fa, jbrowse_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0)\n",
      "INFO: Informed scheduler that task   AddGenome__usr_users_JIC_a__usr_users_JIC_a_5f48dcba3f   has status   DONE\n",
      "INFO: [pid 15281] Worker Worker(salt=476501024, workers=25, host=n128n20.nbicluster, username=buntingd, pid=14220) running   AddGFF3(reference=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/reference/Pst_104E_v13_p_ctg.fa, base_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data, jbrowse_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0)\n",
      "INFO: Tmp dir: /nbi/scratch/buntingd/AddGFF3__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d\n",
      "INFO: [pid 15284] Worker Worker(salt=476501024, workers=25, host=n128n20.nbicluster, username=buntingd, pid=14220) running   AddGFF2(reference=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/reference/Pst_104E_v13_p_ctg.fa, base_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data, jbrowse_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0)\n",
      "INFO: Tmp dir: /nbi/scratch/buntingd/AddGFF2__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d\n",
      "INFO: [pid 15289] Worker Worker(salt=476501024, workers=25, host=n128n20.nbicluster, username=buntingd, pid=14220) running   AddGFF1(reference=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/reference/Pst_104E_v13_p_ctg.fa, base_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data, jbrowse_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0)\n",
      "INFO: Tmp dir: /nbi/scratch/buntingd/AddGFF1__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d\n",
      "INFO: SLURM: jobid=15034885\n",
      "INFO: SLURM: jobid=15034886\n",
      "INFO: SLURM: jobid=15034884\n",
      "INFO: \n",
      "SLURM err AddGFF1__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d: \n",
      "SLURM out AddGFF1__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d: \n",
      "INFO: [pid 15289] Worker Worker(salt=476501024, workers=25, host=n128n20.nbicluster, username=buntingd, pid=14220) done      AddGFF1(reference=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/reference/Pst_104E_v13_p_ctg.fa, base_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data, jbrowse_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0)\n",
      "INFO: Informed scheduler that task   AddGFF1__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d   has status   DONE\n",
      "INFO: \n",
      "SLURM err AddGFF2__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d: \n",
      "SLURM out AddGFF2__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d: \n",
      "INFO: [pid 15284] Worker Worker(salt=476501024, workers=25, host=n128n20.nbicluster, username=buntingd, pid=14220) done      AddGFF2(reference=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/reference/Pst_104E_v13_p_ctg.fa, base_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data, jbrowse_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0)\n",
      "INFO: Informed scheduler that task   AddGFF2__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d   has status   DONE\n",
      "INFO: \n",
      "SLURM err AddGFF3__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d: \n",
      "SLURM out AddGFF3__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d: \n",
      "INFO: [pid 15281] Worker Worker(salt=476501024, workers=25, host=n128n20.nbicluster, username=buntingd, pid=14220) done      AddGFF3(reference=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/reference/Pst_104E_v13_p_ctg.fa, base_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data, jbrowse_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0)\n",
      "INFO: Informed scheduler that task   AddGFF3__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d   has status   DONE\n",
      "INFO: [pid 15503] Worker Worker(salt=476501024, workers=25, host=n128n20.nbicluster, username=buntingd, pid=14220) running   MAKERPipelineWrapper(reference=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/reference/Pst_104E_v13_p_ctg.fa, base_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data, jbrowse_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0)\n",
      "INFO: [pid 15503] Worker Worker(salt=476501024, workers=25, host=n128n20.nbicluster, username=buntingd, pid=14220) done      MAKERPipelineWrapper(reference=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/reference/Pst_104E_v13_p_ctg.fa, base_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data, jbrowse_dir=/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0)\n",
      "INFO: Informed scheduler that task   MAKERPipelineWrapper__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d   has status   DONE\n",
      "INFO: Worker Worker(salt=476501024, workers=25, host=n128n20.nbicluster, username=buntingd, pid=14220) was stopped. Shutting down Keep-Alive thread\n",
      "INFO: \n",
      "===== Luigi Execution Summary =====\n",
      "\n",
      "Scheduled 14 tasks of which:\n",
      "* 9 present dependencies were encountered:\n",
      "    - 1 FathomStats1(...)\n",
      "    - 1 FathomStats2(...)\n",
      "    - 1 GFFMerge1(...)\n",
      "    - 1 GFFMerge2(...)\n",
      "    - 1 GFFMerge3(...)\n",
      "    ...\n",
      "* 5 ran successfully:\n",
      "    - 1 AddGFF1(...)\n",
      "    - 1 AddGFF2(...)\n",
      "    - 1 AddGFF3(...)\n",
      "    - 1 AddGenome(...)\n",
      "    - 1 MAKERPipelineWrapper(...)\n",
      "\n",
      "This progress looks :) because there were no failed tasks or missing external dependencies\n",
      "\n",
      "===== Luigi Execution Summary =====\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "luigi.build([MAKERPipelineWrapper(reference=reference,\n",
    "                    base_dir=\"/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data\",\n",
    "                    jbrowse_dir='/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0')],\n",
    "            workers=25,\n",
    "            local_scheduler=False,\n",
    "            scheduler_host='j512n1',\n",
    "            log_level='INFO')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "print(luigi.tools.deps_tree.print_tree(MAKERPipelineWrapper(reference=reference,\n",
    "                      base_dir=\"/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data\",\n",
    "                      jbrowse_dir='/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AddGFF3__usr_users_JIC_a__usr_users_JIC_a__usr_users_JIC_a_7b759e266d'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AddGFF3(reference=reference,\n",
    "       jbrowse_dir='/usr/users/JIC_a1/buntingd/GenomeAnnotation/Jbrowse/JBrowse-1.15.0',\n",
    "       base_dir=\"/usr/users/JIC_a1/buntingd/GenomeAnnotation/PST104/data\").task_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
